### PR TITLE
Precise Nix logo rework (aka genix7000)

### PR DESCRIPTION
**Pull request by cab404**
This is a Nix logo version, generated in OpenSCAD and pinned to a hex grid. Differences between generated and original versions are neglidgible:

Old and new logos, overlayed

*Blue is a newer version, red is an older version*

Here's a "hexgrid" version, which shows the proportions better:

Nix logo with different weights

## Why OpenSCAD?

Having most of the logo parametrized allows to generate logos with different weights with ease

```openscad
   for (gap=[1:5]) {
        $gaps = gap * 2;
        translate([(gap - 1) * 160,((gap - 1) % 2) * 160])
        difference()
        render_logo();
    };
```
Nix logo with different weights

It also allows to easily animate the logo — [here's an example](https://raw.githubusercontent.com/cab404/nixos-artwork/master/ng/doc/fadeout.mp4) — e.g if you want to have minimalistic plymouth animation.

Moreover, having even our logo configurable and reproducible better suits NIx as a project ♥  

## 3D Printing

nix.scad can also export several 3D printing models — single-module smaller versions, and separate, composable lambdas — if you wish to build yourself a huge Nix logo.
Nix logo module

## Hacking on it

Run `nix-shell -p openscad` if you don't have an OpenSCAD yet, and then open nix.scad in it.

If you wish to export colored logos, modify and use `./export-color-svgs.sh` in this folder.

### CODE REVIEW

**Comment by cab404** [comment]
This originated from a thread on discourse: https://discourse.nixos.org/t/genix7000-nix-project-logo-generator/15937

File: ng/nix.scad [hunk 1]
```diff
@@ -0,0 +1,220 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+// This is a more rigid version of
```
Comment (tazjin) [inline ng/nix.scad:3]: of? :)

**Comment by samueldr** [comment]
HI! :wave:

I really like the idea/concept.

I like the fact that it makes it easier, nay, possible to program animations that play with the logo metrics. I *personally* find the current example awkward, but that's, like, my opinion.

> *Awkward how?*
>
> I think anything that scales the relative weight of the lambdaflake branches will be awkward. In your example since the overall footprint of the hexagon changes, and the lambdaflake branches are constant sized, their relative size is "bolder", wider, when the hexagon footprint is smaller. It's not until the footprint is fully realized that the relative weights are *correct*.
>
> Does that make sense?

* * *

#### How can we move this forward?

The official "proper" logo has shading darkening the lambdaflake branches pointing towards the central hexagon. Do you think it's possible to integrate this into the SVG generation from OpenSCAD? I figure it's possible the gradient itself can't, but we might be able to apply the gradient after the fact, not dissimilar to how you're already manipulating the SVG in a shell script.

Additionally, [one of the colour is not correct according to the guidelines](https://github.com/NixOS/nixos-artwork/tree/master/logo#colours).

Then we could drop the `ng` (really rename it OpenSCAD or something like that). I think we can make this the official logo. It's not like the differences matter. Then we probably want to rework the README into a README about the OpenSCAD model.

The "diff" and "fadeout" outputs, which really are used for this PR can be dropped.

The doc header we'd maybe want to see if we can graduate into a more proper header for official README uses? (Which would require a light theme/dark theme version.)

By dropping the `ng` folder, we'd also drop the `out` folder and bake-in the correct variants in the existing files. 

* * *

#### Some more nits

There are only two canon versions of the logo. The coloured one (blue and light blue), and the "monochrome" one. The monochrome one **has** to be used white on dark background, or "ink colour" in monochrome printouts.

I don't know how, but we should document that the tools are meant to make the logo more correct, and "more reproducible", and to make some programmatic effects possible. That it's not meant to make a fractal of slightly "wrong" logos.

I understand and totally get that some variants like *gapless* and *grid* are helpful for development/design, and to prove correctness. But I'm scared that any variants committed in the repo, or *trivially buildable* will be accidentally used (or willingly used) instead of the correct logo.

I'm saying this since we've also somehow already seen this happen with previous proposals that weren't even committed and accepted in the repo. *Somehow* those variants end up indexed high in image search and people who don't know better use the wrong ones.

**Comment by cab404** [comment]
@samueldr Hello!

Nice to hear feedback from you ♥

### Becoming official logo
Yay!

### Nits
- [ ] Add gradients
- [ ] Add monochrome exports
- [ ] Fix dark color
   Actually I kinda missed that due to there being a gradients — I didn't find the logo docs at the time)

### Alt-logos
Yeah, I guess we can cut them from the main file, and provide only as export options in openscad (as well as marking those options «unofficial». This way I think it would be too high effort for it to be used extensively across the web.

### Weights example
I guess a bit more thought out algo helps.
```openscad
for (gap=[1:5]) {
        $gaps = 1;
        $aperture = 2 * gap;
        $length = 4 * gap;
        $clipr = 8 * gap;
        $unit = 20 / gap;
        translate([(gap - 1) * 160,((gap - 1) % 2) * 160])
        render_logo();
    };
```
test


In any case, having different weights for the logo is a bit out of scope for the goal of creating a replacement for two current logos, alas an interesting possibility this model provides.

**Comment by cab404** [comment]
image
Okay, I think my fc shim started working, and now I am able to specify correct fonts to OpenSCAD without changing system state)

**Review by dasJ — APPROVED** [review]

**Review by mweinelt — APPROVED** [review]

**Comment by cab404** [comment]
I would love some help with programmatically applying gradients — if anyone wants to take a shot at it, you are most welcome!
If you would need my help, feel free to contact me and schedule a call)

**Comment by blaggacao** [comment]
```js
// SPDX-License-Identifier: GPL-3.0-or-later

// The angle™
function th(params) {
  return 360 / params.num / 2;
}

// Unit value of Y when mapped to coordinate space with angle between axes of "th"
function tunit(params) {
  return Math.tan(th(params) * (Math.PI / 180)) * params.unit;
}

// === Nix logo specification

const params = {
  // Central aperture diameter, in units. It does produce nice effects if animated.
  aperture: 2,

  // lambda height in units. fun to play with
  length: 4,

  // Clipping polygon diameter, in units
  clipr: 7,

  // number of lambdas. doesn't really work if changed in this model.
  num: 6,

  // Lambda thickness, also a segment size.
  // Should affect nothing except size and gaps.
  unit: 25,

  // Shrinkage for each of lambdas. Basically control inverse "font weight"
  gaps: 1,

  // colors to use
  colors: ["#5277C3", "#7EBAE4"],
};

Snap.plugin(function (Snap, Element, Paper, global) {
  Paper.prototype.cRect = function (x, y) {
    return this.rect(-x / 2, -y / 2, x, y);
  };
  Paper.prototype.cUse = function (id) {
    return this.use(id).attr({ x: -500, y: -500 });
  };
});

var s = Snap("#out").attr({
  viewBox: "-500 -500 1000 1000",
});

// Draw the lambda symbol
var lambda = s.symbol(-500, -500, 1000, 1000);
{
  let ax = params.unit;
  let ay = params.length * tunit(params) * 2;
  let a = s.cRect(ax, ay).transform(
    Snap.matrix()
      .rotate(th(params))
      .translate(0, ay / 2)
  );

  let bx = ax;
  let by = tunit(params) * (params.length * 2 + 2);
  let b = s.cRect(bx, by).transform(Snap.matrix().rotate(-th(params)));

  let cx = tunit(params) * (params.length + 2);
  let cy = params.unit * params.length;
  let c = s.cRect(cx, cy).attr({ fill: "white" });

  // mask the next one from the bottom of the lambda's leg
  let n = s.cRect(bx, by).transform(
    Snap.matrix()
      .rotate(-th(params))
      .translate(tunit(params) * params.aperture, params.unit * params.aperture)
      .invert()
  );

  let mask = s.mask().toDefs();

  let g = s.group(a, b);

  mask.add(c);
  mask.add(n);
  g.attr({ mask });

  // offset each lambda
  g.transform(Snap.matrix().scale(1 - params.gaps / 10));

  lambda.add(g);
}

var lambdas = [];

function render() {
  for (let i = 0; i < params.num; i++) {
    let color = params.colors[i % params.colors.length];
    let b = s
      .cUse(lambda)
      .attr({ fill: color })
      .transform(
        Snap.matrix()
          .rotate(-th(params) * 2 * i)
          .translate(
            -tunit(params) * params.aperture,
            params.unit * params.aperture
          )
      );
    lambdas.push(b);
  }
}

let polygonMask = s.mask().toDefs();

// mask the final result with a regular polygon
{
  // copied from <MCAD/regular_shapes.scad> so customizer will work on thingiverse
  function regular_polygon(sides, radius) {
    console.log(sides);
    console.log(radius);
    let angles = [...Array(sides).keys()].map((i) => i * (360 / sides));
    console.log(angles);
    let coords = angles.map((a) => [
      radius * Math.cos(a * (Math.PI / 180)),
      radius * Math.sin(a * (Math.PI / 180)),
    ]);
    console.log(coords);
    return coords;
    s.polygon(coords);
  }

  let p = s
    .polygon(regular_polygon(params.num, params.clipr * tunit(params)))
    .attr({ fill: "white" });

  polygonMask.add(p);
}

render();

s.group(...lambdas).attr({ mask: polygonMask });
```

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <title>Nix Logo Generator</title>
    <style media="screen">
      html,
      body {
        padding: 0;
        margin: 0;
        background: #333;
        height: 100%;
        width: 100%;
      }
      svg {
        display: block;
      }
    </style>
    <script
      src="https://cdnjs.cloudflare.com/ajax/libs/snap.svg/0.5.1/snap.svg-min.js"
      integrity="sha512-Gk+uNk8NWN235mIkS6B7/424TsDuPDaoAsUekJCKTWLKP6wlaPv+PBGfO7dbvZeibVPGW+mYidz0vL0XaWwz4w=="
      crossorigin="anonymous"
      referrerpolicy="no-referrer"
    ></script>

  </head>
  <body>
    <svg id="out" xmlns="http://www.w3.org/2000/svg" />
    <script src="./nix-logo.js"></script>
  </body>
</html>
```

image


---
so far so good ...

**Comment by blaggacao** [comment]
Here is a complete example: the logo downloads as svg on click...

https://github.com/divnix/std/pull/184

**Comment by cab404** [comment]
> Here is a complete example: the logo downloads as svg on click...
> 
> [divnix/std#184](https://github.com/divnix/std/pull/184)

ooo, nice

can you apply gradients?

**Comment by cab404** [comment]
I guess I'll try doing it with snap then! <3

**Comment by blaggacao** [comment]
I fixed the width a little later, it should be good (or better) now.

I think gradients can be applied. I've tried to work with fonts, but embbeding fonts is a bit of a nightmare, I'll probably need to find a path tracer lib, or something.

Maybe it's also possible to post-process with `inkscape`.

This was really just a first attempt and it took me a while to get the origins/offsets right. Still not sure, my chosen path (via symbols) is the best way, there seem to be some edge cases w.r.t. masking.

**Comment by cab404** [comment]
> Maybe it's also possible to post-process with `inkscape`.

I've tried — gradients are not exposed in the shell :(

**Comment by blaggacao** [comment]
Looka like this might be good enough to get text to path post-processing done (for logos with text that work): https://github.com/paulzi/svg-text-to-path

**Review by MatthewCroughan — APPROVED** [review]

**Review by djacu — APPROVED** [review]

**Pull request merged**